File: fe/src/utils/subscribeUser.ts
const CLOUD189_SUBSCRIBE_USER_LINK_RE =
  /(?:^|[^a-z0-9_+./:@-])(?:(?:https?:\/\/|\/\/)?)content\.21cn\.com(?::\d+)?(?:\/[^\s"'<>]*)?[?&]uuid=([^&#\s"'<>]+)/i
const CLOUD189_SUBSCRIBE_USER_LABEL_RE =
  /(?:订阅号(?:ID)?|订阅用户(?:ID)?|\bsubscribe[\s_-]?user(?:id)?\b|\bup[\s_-]?user(?:id)?\b|\buuid\b)[:：=]\s*([^\s,，;；。()（）]+)/i
const CLOUD189_SUBSCRIBE_USER_BOUNDARY_TRIM_RE =
  /^[\s.,，。;；:：!?！？、()[\]【】<>《》「」『』“”‘’"']+|[\s.,，。;；:：!?！？、()[\]【】<>《》「」『』“”‘’"']+$/g
const CLOUD189_URL_LIKE_RE =
  /(?:[a-z][a-z0-9+.-]*:\/\/|\/\/)[^\s"'<>]+|(?:^|[^a-z0-9_+./:@-])(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+[a-z]{2,}(?::\d+)?(?:\/[^\s"'<>]*|\?[^\s"'<>]*)/i

export const parseCloud189SubscribeUserId = (input: string) => {
  let match = input.match(CLOUD189_SUBSCRIBE_USER_LINK_RE)
  if (match?.[1]) {
    return normalizeSubscribeUserCandidate(match[1])
  }

  if (CLOUD189_URL_LIKE_RE.test(input)) {
    return ''
  }

  match = input.match(CLOUD189_SUBSCRIBE_USER_LABEL_RE)
  if (match?.[1]) {
    return normalizeSubscribeUserCandidate(match[1])
  }

  return ''
}

export const normalizeCloud189SubscribeUserInput = (input: string) => {
  const parsed = parseCloud189SubscribeUserId(input)
  if (parsed) {
    return parsed
  }

  input = trimSubscribeUserBoundary(input)
  if (CLOUD189_URL_LIKE_RE.test(input)) {
    return ''
  }

  return input
}

const normalizeSubscribeUserCandidate = (input: string) => {
  let subscribeUserId = trimSubscribeUserBoundary(input)
  if (!subscribeUserId) {
    return ''
  }

  try {
    subscribeUserId = trimSubscribeUserBoundary(
      decodeURIComponent(subscribeUserId.replace(/\+/g, ' '))
    )
  } catch {
    return subscribeUserId.trim()
  }

  return subscribeUserId.trim()
}

const trimSubscribeUserBoundary = (value: string) => {
  return value.trim().replace(CLOUD189_SUBSCRIBE_USER_BOUNDARY_TRIM_RE, '')
}

File: internal/pkgs/utils/subscribe_user.go
package utils

import (
	"net/url"
	"regexp"
	"strings"
)

var cloud189SubscribeUserLinkPattern = regexp.MustCompile(
	`(?i)(?:^|[^a-z0-9_+./:@-])(?:(?:https?://|//)?)content\.21cn\.com(?::\d+)?(?:/[^\s"'<>]*)?[?&]uuid=([^&#\s"'<>]+)`,
)

var cloud189SubscribeUserLabelPattern = regexp.MustCompile(
	`(?i)(?:订阅号(?:ID)?|订阅用户(?:ID)?|\bsubscribe[\s_-]?user(?:id)?\b|\bup[\s_-]?user(?:id)?\b|\buuid\b)[:：=]\s*([^\s,，;；。()（）]+)`,
)

const cloud189SubscribeUserBoundaryTrimChars = " \t\r\n.,，。;；:：!?！？、()[]【】<>《》「」『』“”‘’\"'"

func ParseCloud189SubscribeUserID(input string) string {
	matches := cloud189SubscribeUserLinkPattern.FindStringSubmatch(input)
	if len(matches) > 1 {
		return normalizeCloud189SubscribeUserCandidate(matches[1])
	}

	if ContainsURLLike(input) {
		return ""
	}

	matches = cloud189SubscribeUserLabelPattern.FindStringSubmatch(input)
	if len(matches) > 1 {
		return normalizeCloud189SubscribeUserCandidate(matches[1])
	}

	return ""
}

func normalizeCloud189SubscribeUserCandidate(input string) string {
	subscribeUserID := strings.Trim(input, cloud189SubscribeUserBoundaryTrimChars)
	if subscribeUserID == "" {
		return ""
	}

	if decoded, err := url.QueryUnescape(subscribeUserID); err == nil {
		subscribeUserID = strings.Trim(decoded, cloud189SubscribeUserBoundaryTrimChars)
	}

	return strings.TrimSpace(subscribeUserID)
}

func NormalizeCloud189SubscribeUserID(input string) string {
	if parsed := ParseCloud189SubscribeUserID(input); parsed != "" {
		return parsed
	}

	input = strings.Trim(input, cloud189SubscribeUserBoundaryTrimChars)
	if ContainsURLLike(input) {
		return ""
	}

	return input
}
